File: patternlets/MPI/11.scatter/scatter.c
/* scatter.c
 * ... illustrates the use of MPI_Scatter()...
 * Joel Adams, Calvin College, November 2009.
 *
 * Usage: mpirun -np N ./scatter
 *
 * Exercise:
 * - Compile and run, varying N: 1, 2, 4, 8
 * - Trace execution through source code.
 * - Explain behavior/effect of MPI_Scatter().
 */

#include <mpi.h>      // MPI
#include <stdio.h>    // printf(), etc.
#include <stdlib.h>   // malloc()

void print(int id, char* arrName, int* arr, int arrSize);

int main(int argc, char** argv) {
    const int MAX = 8;
    int* arrSend = NULL;
    int* arrRcv = NULL;
    int numProcs = -1, myRank = -1, numSent = -1;

    MPI_Init(&argc, &argv);                            // initialize
    MPI_Comm_size(MPI_COMM_WORLD, &numProcs);
    MPI_Comm_rank(MPI_COMM_WORLD, &myRank);

    if (myRank == 0) {                                 // master process:
        arrSend = (int*) malloc( MAX * sizeof(int) );  //  allocate array1
        for (int i = 0; i < MAX; i++) {                //  load with values
            arrSend[i] = (i+1) * 11;
        }
        print(myRank, "arrSend", arrSend, MAX);        //  display array1
    }
     
    numSent = MAX / numProcs;                          // all processes:
    arrRcv = (int*) malloc( numSent * sizeof(int) );   //  allocate array2

    MPI_Scatter(arrSend, numSent, MPI_INT, arrRcv,     //  scatter array1 
                 numSent, MPI_INT, 0, MPI_COMM_WORLD); //   into array2

    print(myRank, "arrRcv", arrRcv, numSent);          // display array2

    free(arrSend);                                     // clean up
    free(arrRcv);
    MPI_Finalize();
    return 0;
}

void print(int id, char* arrName, int* arr, int arrSize) {
    printf("Process %d, %s: ", id, arrName);
    for (int i = 0; i < arrSize; i++) {
        printf(" %d", arr[i]);
    }
    printf("\n");
}

File: patternlets/java-OpenMPI/15.scatter/Scatter.java
/* Scatter.java
 * ... illustrates the basic scatter pattern 
 *      using buffers in OpenMPI's Java interface.
 *
 * Note: If the number of values being scattered is not
 *       evenly divisible by the number of processes,
 *       use scatterv() instead of scatter.
 *
 * Joel Adams, Calvin University, November 2019.
 *
 * Usage: mpirun -np 4 java ./Scatter
 *
 * Exercise:
 * - Compile, then run with N = 1, 2, 4, 8. 
 * - Trace execution through source code. 
 * - Explain behavior/effect of scatter. 
 * - What if BUFFER_SIZE is not evenly divisible by N?
 */

import mpi.*;
import java.nio.IntBuffer;

public class Scatter {

 public static void main(String [] args) throws MPIException {
    MPI.Init(args);

    Comm comm         = MPI.COMM_WORLD;
    int id            = comm.getRank();
    int numProcesses  = comm.getSize();

    if (numProcesses > BUFFER_SIZE) {
        if (id == MASTER) {
            System.out.println("\nPlease run this program with N <= 8 processes\n");
        }
        MPI.Finalize();
        System.exit(0);
    }
    IntBuffer sendBuf = null;
    IntBuffer recvBuf = null;

    if (id == MASTER) {
        sendBuf = MPI.newIntBuffer(BUFFER_SIZE);
        for (int i = 0; i < BUFFER_SIZE; ++i) {
           sendBuf.put(i, (i+1) * 11);
        }
        System.out.print("\nBefore scatter: ");
        printBuf(id, "sendBuf", sendBuf);
    }
 
    int numSent = BUFFER_SIZE / numProcesses;
       
    comm.barrier();                    // see comment on next barrier

    recvBuf = MPI.newIntBuffer(numSent);
    printBuf(id, "recvBuf", recvBuf);

    printSeparator("----", id);

    comm.scatter(sendBuf, numSent, MPI.INT, 
                  recvBuf, numSent, MPI.INT, MASTER); 

    if (id == MASTER) {
        System.out.print("After scatter:\n");
    }
    comm.barrier();                    // all of these barriers are here
    printBuf(id, "recvBuf", recvBuf);  //  just to make the output easier
    comm.barrier();                    //  to read; no effect on correctness
    if (id == MASTER) {
        System.out.println();
    }

    MPI.Finalize();
  }

  /* utility to display the contents of an IntBuffer.
   * @param: id, the int MPI rank of this process.
   * @param: bufName, a String that is the name of the buffer.
   * @param: buf, the IntBuffer.
   * @param: size, the size of buf.
   */
  private static void printBuf(int id, String bufName, IntBuffer buf) {
      String msg = "Process " + id + ", " + bufName + ": [";
      int size = buf.capacity();
      int sizeLessOne = size - 1;
      for (int i = 0; i < size; ++i) {
         msg += buf.get(i);
         if (i < sizeLessOne ) {
             msg += ",";
         }
      }
      msg += "]\n";
      System.out.print(msg);
  }

  /* utility to print a separator between before and after sections.
   * @param: separator, a String.
   * @param: id, the MPI rank of this process. 
   * POST: separator has been printed by the master process.
   */
  private static void printSeparator(String separator, int id) throws MPIException {
     MPI.COMM_WORLD.barrier();
     if (id == MASTER) { System.out.print(separator + "\n"); }
     MPI.COMM_WORLD.barrier();
  }

  private static int BUFFER_SIZE = 8;
  private static int MASTER      = 0;
}

